packages/app/src/cli/services/function/replay.ts

```diff
@@ -46,17 +49,39 @@ export interface FunctionRunData {
 }
 
 export async function replay(options: ReplayOptions) {
-  const {apiKey} = await ensureConnectedAppFunctionContext(options)
-  const functionRunsDir = joinPath(getLogsDir(), apiKey)
+  const {watch, extension, app} = options
+  const abortController = new AbortController()
 
-  const functionRuns = await getFunctionRunData(functionRunsDir, options.extension.handle)
-  const selectedRun = await selectFunctionRunPrompt(functionRuns)
+  try {
+    const {apiKey} = await ensureConnectedAppFunctionContext(options)
+    const functionRunsDir = joinPath(getLogsDir(), apiKey)
+    const functionRuns = await getFunctionRunData(functionRunsDir, options.extension.handle)
 
-  if (selectedRun === undefined) {
-    throw new AbortError(`No logs found in ${functionRunsDir}`)
-  }
+    let selectedRun = await selectFunctionRunPrompt(functionRuns)
+
+    if (selectedRun === undefined) {
+      throw new AbortError(`No logs found in ${functionRunsDir}`)
+    }
 
-  await runFunctionRunnerWithLogInput(options.extension, options, selectedRun.payload.input, selectedRun.payload.export)
+    const {input, export: runExport} = selectedRun.payload
+    await runFunctionRunnerWithLogInput(extension, options, input, runExport)
+
+    if (watch) {
+      await setupExtensionWatcher({
+        extension,
+        app,
+        stdout: process.stdout,
+        stderr: process.stderr,
+        onChange: async () => {
+          await runFunctionRunnerWithLogInput(extension, options, input, runExport)
+        },
+        signal: abortController.signal,
+      })
+    }
+  } catch (e) {
+    abortController.abort()
```

Comment:
We pass in an abort controller so that the underlying code can clean up if we need to abort. Similar to how we handle errors in `dev` [here](https://github.com/Shopify/cli/blob/main/packages/app/src/cli/services/dev/ui/components/Dev.tsx#L97), I thought it made sense to abort if we catch an error at this level.